### ISSUE TITLE
fly.io is closing connections due to hard limit

### ISSUE THREAD

**Issue by bard** [gitcoinco/grants-stack-indexer#153]
Connections might be hanging:

```
2023-07-13T14:34:04.378 proxy[5683776a714d8e] den [warn] Instance 5683776a714d8e reached hard
limit of 25 concurrent connections. This usually indicates your app is not closing connections properly 
or is not closing them fast enough for the traffic levels it is handling. Scaling resources, number of 
instances or increasing your hard limit might help.
```

**Comment by dkotik** [comment]
I have the same problem on production service, even though the config has really high concurrency limits:

```toml
  [services.concurrency]
    type = "connections"
    hard_limit = 9999999999
    soft_limit = 99999
```

**Comment by dkotik** [comment]
According to [this ticket](https://community.fly.io/t/problem-increasing-apps-connections-hard-limit/13913/2), instead of `services` the config should use `http_services`

**Comment by dkotik** [comment]
Although, that can't be right, because when I run

```sh
fly m status 1781329f523008  -d | grep hard_limit
```

it gives me correct super-huge hard limit instead of 25

**Comment by bard** [comment]
Closing as this hasn't reappeared so it's hopefully fixed on Fly's side.

**Comment by dkotik** [comment]
I am still having the same issue on a small instance. I changed concurrency to "requests" and increased instance size to shared CPU 2x. Posting here, because this page still comes up third on google search for this problem.

**Comment by bard** [comment]
Reopening this as it's occurring again. 25 concurrent requests might be too low, in that case [it can be increased](https://fly.io/docs/reference/load-balancing/#web-service). /cc @boudra

**Comment by boudra** [comment]
I was checking the app concurrency metric in Grafana to see how this happens, I noticed the way Fly balances connections is weird, because at some point you can see two instances serving around 20 requests and there's one that is serving 0.

I see the CPU usage is mostly averaging 50% with occasional spikes to 80%, I will add another instance just in case and increase the hard limit once it is live, maybe set the hard limit to 40 and see what happens?

**Comment by boudra** [comment]
I am wondering whether this hard limit is also applying to the health check requests and not just external requests.

**Comment by bard** [comment]
> I see the CPU usage is mostly averaging 50% with occasional spikes to 80%, but request times are high (for some reason directory listings are slower)

Interesting, do I understand correctly that hitting e.g. https://indexer-staging.fly.dev/data/1/projects.json is consistently faster than hitting https://indexer-staging.fly.dev/data/?

> Also, I assume when connections are dropped Fly returns a 5xx? I don't see a lot of them in the metrics, maybe they just close the connection?

The docs say: "If traffic is far above the hard_limit for a long period of time, Fly.io may start returning 503 Service Unavailable responses for requests that are not able to be routed to an instance."

> I am wondering whether this hard limit is also applied to the health check requests and not just external requests, as soon as I see that warning, the health check starts failing, it makes sense that they would happen at the same time and one might not cause the other but it's suspicious 🤔

The limit _shouldn't_ apply to the health check request since the health check request is supposed to hit the machine directly without going through the load balancer (if it did, it could be routed to a different machine). Maybe the health check is failing first, which causes 25 "slots" to suddenly be removed from the pool, thereby hitting the hard limit?

**Comment by bard** [comment]
Also seeing this in [logs](https://app.datadoghq.eu/logs?query=%40fly.app.name%3A%22indexer-production%22%20host%3Aed8c&cols=host%2Cservice&event=AgAAAYvo17aUiHDKewAAAAAAAAAYAAAAAEFZdm8xN2FVQUFCbzIxNVRFakE4WmdBaQAAACQAAAAAMDE4YmU4ZGEtMWQ2ZS00ZWExLWJmYmYtMWM4ODM5N2NjZmIx&index=%2A&messageDisplay=inline&refresh_mode=sliding&stream_sort=time%2Cdesc&view=spans&viz=stream&from_ts=1693999097397&to_ts=1694085497397&live=true), maybe related:

```
(node:346) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added to [TLSSocket]. Use emitter.setMaxListeners() to increase limit
```

One confusing thing is that the `host` field does not match the `instance` reported in the log message:

![image](https://github.com/gitcoinco/grants-stack-indexer/assets/19322/ff8d284e-27bd-4bfe-92c1-ac97eed93b3e)


O

**Comment by boudra** [comment]
> Interesting, do I understand correctly that hitting e.g. https://indexer-staging.fly.dev/data/1/projects.json is consistently faster than hitting https://indexer-staging.fly.dev/data/?

Yes, it's very weird, it's not happening to me anymore though, I remember it being very noticeable, files loading instantly and directory listing taking up to 5 seconds.

> The limit shouldn't apply to the health check request since the health check request is supposed to hit the machine directly without going through the load balancer (if it did, it could be routed to a different machine). Maybe the health check is failing first, which causes 25 "slots" to suddenly be removed from the pool, thereby hitting the hard limit?

Yeah that makes a lot of sense, I've also been seeing hard limit warnings without health checks failing lately so definitely not related.

> (node:346) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added to [TLSSocket]. Use emitter.setMaxListeners() to increase limit

Not sure where this error is happening, could be in Chainsauce, we're running the version with recursive RPC fetching in production. Opened an issue here: https://github.com/gitcoinco/grants-stack-indexer/issues/361
